File: mobile/src/geolocate.ts
/**
 * geolocate.ts — resolve the device's real position + an Azerbaijani address.
 *
 * Returns { lat, lng, location }. The `location` string is reverse-geocoded via
 * OpenStreetMap's Nominatim; if that fails it falls back to a coordinate label.
 * Requires a secure context (HTTPS or localhost) for navigator.geolocation.
 */

export interface ResolvedLocation {
  lat: number;
  lng: number;
  location: string;
}

function coordLabel(lat: number, lng: number): string {
  return `${lat.toFixed(4)}°N, ${lng.toFixed(4)}°E`;
}

async function reverseGeocode(lat: number, lng: number): Promise<string> {
  try {
    const res = await fetch(
      `https://nominatim.openstreetmap.org/reverse?lat=${lat}&lon=${lng}&format=json&zoom=18&accept-language=az`,
    );
    if (!res.ok) return coordLabel(lat, lng);
    const data = await res.json();
    const a = data.address ?? {};
    const road = a.road || a.pedestrian || a.footway || a.residential || '';
    const area = a.suburb || a.neighbourhood || a.city_district || a.quarter || a.city || '';
    const parts = [road, area].filter(Boolean);
    return parts.length ? parts.join(', ') : coordLabel(lat, lng);
  } catch {
    return coordLabel(lat, lng);
  }
}

/** Resolve current position. Rejects if geolocation is unavailable or denied. */
export function resolveLocation(): Promise<ResolvedLocation> {
  return new Promise((resolve, reject) => {
    if (!navigator.geolocation) {
      reject(new Error('Geolocation unavailable'));
      return;
    }
    navigator.geolocation.getCurrentPosition(
      async pos => {
        const { latitude: lat, longitude: lng } = pos.coords;
        const location = await reverseGeocode(lat, lng);
        resolve({ lat, lng, location });
      },
      err => reject(err),
      { enableHighAccuracy: true, timeout: 12000, maximumAge: 30000 },
    );
  });
}

File: mobile_flutter/lib/geolocate.dart
/// geolocate.dart — resolve the device's real position + an Azerbaijani address.
/// Ported from mobile/src/geolocate.ts (browser Nominatim → native geocoding).
library;

import 'package:geolocator/geolocator.dart';
import 'package:geocoding/geocoding.dart';

class ResolvedLocation {
  final double lat;
  final double lng;
  final String location;
  ResolvedLocation(this.lat, this.lng, this.location);
}

String _coordLabel(double lat, double lng) =>
    '${lat.toStringAsFixed(4)}°N, ${lng.toStringAsFixed(4)}°E';

Future<String> _reverseGeocode(double lat, double lng) async {
  try {
    final placemarks = await placemarkFromCoordinates(lat, lng);
    if (placemarks.isEmpty) return _coordLabel(lat, lng);
    final p = placemarks.first;
    final road = (p.thoroughfare?.isNotEmpty ?? false)
        ? p.thoroughfare!
        : (p.street ?? '');
    final area = (p.subLocality?.isNotEmpty ?? false)
        ? p.subLocality!
        : (p.locality ?? '');
    final parts = [road, area].where((s) => s.isNotEmpty).toList();
    return parts.isEmpty ? _coordLabel(lat, lng) : parts.join(', ');
  } catch (_) {
    return _coordLabel(lat, lng);
  }
}

/// Resolve current position. Throws if location services/permission unavailable.
Future<ResolvedLocation> resolveLocation() async {
  final enabled = await Geolocator.isLocationServiceEnabled();
  if (!enabled) throw Exception('Location services disabled');

  var permission = await Geolocator.checkPermission();
  if (permission == LocationPermission.denied) {
    permission = await Geolocator.requestPermission();
  }
  if (permission == LocationPermission.denied ||
      permission == LocationPermission.deniedForever) {
    throw Exception('Location permission denied');
  }

  final pos = await Geolocator.getCurrentPosition(
    locationSettings: const LocationSettings(accuracy: LocationAccuracy.high),
  );
  final label = await _reverseGeocode(pos.latitude, pos.longitude);
  return ResolvedLocation(pos.latitude, pos.longitude, label);
}
